cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -7,15 +7,33 @@ private import DataFlowDispatch
  * A data flow node that occurs as the argument of a call and is passed as-is
  * to the callable. Instance arguments (`this` pointer) are also included.
  */
-class ArgumentNode extends InstructionNode {
+class ArgumentNode extends Node {
   ArgumentNode() {
-    exists(CallInstruction call |
-      instr = call.getAnArgument()
-      or
-      instr.(ReadSideEffectInstruction).getPrimaryInstruction() = call
-    )
+    // To avoid making this class abstract, we enumerate its values here.
+    this instanceof PrimaryArgumentNode or
+    this instanceof SideEffectArgumentNode
+  }
+
+  /**
+   * Holds if this argument occurs at the given position in the given call.
+   * The instance argument is considered to have index `-1`.
+   */
+  predicate argumentOf(DataFlowCall call, int pos) {
+    this.(PrimaryArgumentNode).argumentOf(call, pos) or
+    this.(SideEffectArgumentNode).argumentOf(call, pos)
   }
 
+  /** Gets the call in which this node is an argument. */
+  DataFlowCall getCall() { this.argumentOf(result, _) }
+}
+
+/**
+ * A data flow node that occurs as the argument to a call, or an
+ * implicit `this` pointer argument.
+ */
+private class PrimaryArgumentNode extends InstructionNode {
```

Comment:
I think ideally this class would be an `OperandNode` representing a `ArgumentOperand`. Why should we represent the argument by the instruction that generates it when we now have a way to refer to the argument itself? Using `ArgumentOperand` would also mean that if `isSource` or `isAdditionalFlowStep` refers to an `ArgumentOperand` then there will be interprocedural flow from it.

It's not as problematic here as [it would have been](https://github.com/github/codeql/pull/3123#discussion_r433916902) in `SideEffectArgumentNode`, though: I believe that an argument is always a register instruction and that a register instruction never has more than one use.

Reply:
I didn't make this class an `OperandNode` to keep the diff minimal. Your comment about `isSource` and `isAdditionalFlowStep` is a good argument for why we should just do it in this PR, though. So I'll do that!

(Register instructions can have more than one use in some situations. I think I've only seen it in constructor and aggregate initializers.)

Reply:
Done in 6c87b08c69d10ecebc2889c5d4c1c4ffa09518f3.